### PR TITLE
DATAUP-497: job message cleanup II

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

More misc housekeeping. Comments inline.

# Jira Ticket / Issue #

Related Jira ticket: https://kbase-jira.atlassian.net/browse/DATAUP-X
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [ ] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (Python) I have run Black and Flake8 on changed Python code manually or with a git precommit hook

### CODE REVIEW

File: src/biokbase/narrative/jobs/jobcomm.py [hunk 1]
```diff
@@ -204,7 +200,7 @@ def __init__(self):
             self._msg_map = {
                 CANCEL: self._cancel_jobs,
                 CELL_JOB_STATUS: self._lookup_job_states_by_cell_id,
-                INFO: self._lookup_job_info,
+                INFO: self._get_job_info,
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobcomm.py:203]: Renamed this to be inline with the JobManager function, which I also renamed.

File: src/biokbase/narrative/jobs/jobcomm.py [hunk 2]
```diff
@@ -392,18 +388,18 @@ def _cancel_jobs(self, req: JobRequest) -> dict:
...
             for result in retry_results.values()
             if "retry" in result
         ]
-
         if len(retry_ids):
             self.send_comm_message(
                 NEW,
                 {JOB_ID_LIST: retry_ids},
             )
+        self.send_comm_message(RETRY, retry_results)
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobcomm.py:401]: Switch order of NEW and RETRY messages -- the new message comes first, followed by the RETRY message.

File: src/biokbase/narrative/jobs/jobmanager.py [hunk 3]
```diff
@@ -268,7 +268,7 @@ def _construct_job_output_state_set(
 
         return output_states
 
-    def lookup_job_info(self, job_ids: List[str]) -> dict:
+    def get_job_info(self, job_ids: List[str]) -> dict:
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobmanager.py:271]: renamed to be more in line with other similar functions -- e.g. `_get_job_ids_by_cell_id` or `_get_jog_logs`

File: src/biokbase/narrative/jobs/jobmanager.py [hunk 4]
```diff
@@ -292,12 +292,7 @@ def lookup_job_info(self, job_ids: List[str]) -> dict:
                 "job_id": job_id,
                 "job_params": job.params,
             }
-        for error_id in error_ids:
-            infos[error_id] = {
-                "job_id": error_id,
-                "error": DOES_NOT_EXIST,
-            }
-        return infos
+        return self.add_errors_to_results(infos, error_ids)
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobmanager.py:295]: uniform handling of the 'job not found' error messages, regardless of what the request is (job status, info, retries, etc.)

File: src/biokbase/narrative/jobs/jobmanager.py [hunk 5]
```diff
@@ -539,20 +524,23 @@ def retry_jobs(self, job_id_list: List[str]) -> dict:
...
+        return self.add_errors_to_results(results_by_job_id, error_ids)
+
+    def add_errors_to_results(self, results: dict, error_ids: List[str]) -> dict:
+        """
+        Add the generic "not found" error for each job_id in error_ids
+        """
+        for error_id in error_ids:
+            results[error_id] = {
+                "job_id": error_id,
+                "error": f"Cannot find job with ID {error_id}",
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobmanager.py:536]: There was a mishmash of ways to handle not found IDs, so to save from having to implement different code for each in the FE, I've standardised them all to this structure.

File: src/biokbase/narrative/tests/job_test_constants.py [hunk 6]
```diff
@@ -1,16 +1,53 @@
 from .util import ConfigTests
 import copy
 
-CLIENTS = "biokbase.narrative.clients.get"
 
 config = ConfigTests()
-TEST_JOBS = config.load_json_file(config.get("jobs", "ee2_job_info_file"))
+TEST_JOBS = config.load_json_file(config.get("jobs", "ee2_job_test_data_file"))
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/job_test_constants.py:6]: file reference renamed to match the file name

File: src/biokbase/narrative/tests/job_test_constants.py [hunk 7]
```diff
@@ -1,16 +1,53 @@
...
+    # collect retried jobs
+    retried_jobs = {}
+    for job in all_jobs.values():
+        # save the first retry ID with the retried job
+        if "retry_ids" in job and len(job["retry_ids"]) > 0:
+            retried_jobs[job["job_id"]] = job["retry_ids"][0]
+    return retried_jobs
+
+
+def generate_error(job_id, err_type):
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/job_test_constants.py:19]: The BE tests deal with various different error messages; these are all valid messages that may get passed to the frontend as part of a result set.

File: src/biokbase/narrative/tests/narrative_mock/mockclients.py [hunk 8]
```diff
@@ -76,33 +82,20 @@ def test_my_function(self):
     """
 
     config = ConfigTests()
-    _ee2_job_info = config.load_json_file(config.get("jobs", "ee2_job_info_file"))
+    _job_state_data = TEST_JOBS
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/narrative_mock/mockclients.py:85]: calling the variable `_ee2_job_info` was confusing as the job info command does something completely different => renamed to `job_state_data`

File: src/biokbase/narrative/tests/narrative_mock/mockclients.py [hunk 9]
```diff
@@ -76,33 +82,20 @@ def test_my_function(self):
...
-    def ee2_job_info(self):
-        return copy.deepcopy(self._ee2_job_info)
-
-    # ----- User and Job State functions -----
-
-    def list_jobs2(self, params):
-        return self.job_info.get("job_info")
-
-    def delete_job(self, job):
-        return "bar"
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/narrative_mock/mockclients.py:99]: not used

File: src/biokbase/narrative/tests/narrative_mock/mockclients.py [hunk 10]
```diff
@@ -76,33 +82,20 @@ def test_my_function(self):
...
+    def job_state_data(self):
+        return copy.deepcopy(self._job_state_data)
 
-    def check_workspace_jobs(self, params):
-        info = self.ee2_job_info
-        if params.get("return_list"):
-            info = list(info.values())
-        return info
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/narrative_mock/mockclients.py:105]: this should be in the ee2 job section

File: src/biokbase/narrative/tests/narrative_mock/mockclients.py [hunk 11]
```diff
@@ -113,7 +106,9 @@ def list_categories(self, params):
         return self.config.load_json_file(self.config.get("specs", "type_specs_file"))
 
     def get_method_full_info(self, params):
-        return self.config.load_json_file(self.config.get("specs", "app_infos_file"))
+        return self.config.load_json_file(
+            self.config.get("specs", "app_full_infos_file")
+        )
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/narrative_mock/mockclients.py:111]: rename file reference to match the file name

File: src/biokbase/narrative/tests/narrative_mock/mockclients.py [hunk 12]
```diff
@@ -268,25 +266,29 @@ def cancel_job(self, params):
             raise generate_ee2_error(CANCEL)
         return {}
 
-    def retry_job(self, params):
-        job_id = params["job_id"]
-        return {"job_id": job_id, "retry_id": job_id[::-1]}
-
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/narrative_mock/mockclients.py:274]: we only ever use the `retry_jobs` endpoint

File: src/biokbase/narrative/tests/narrative_mock/mockclients.py [hunk 13]
```diff
@@ -268,25 +266,29 @@ def cancel_job(self, params):
...
-            results.append({"job_id": job_id, "retry_id": job_id[::-1]})
+            output = {"job_id": job_id}
+            if job_id in RETRIED_JOBS:
+                output["retry_id"] = RETRIED_JOBS[job_id]
+            elif job_id == BATCH_PARENT:
+                output["error"] = generate_error(job_id, "batch_parent_retry")
+            else:
+                output["error"] = generate_error(job_id, "retry_status")
+
+            results.append(output)
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/narrative_mock/mockclients.py:281]: return realistic responses depending on what the input is

File: src/biokbase/narrative/tests/test_job.py [hunk 14]
```diff
@@ -51,17 +56,13 @@ def capture_stdout():
...
 
 
-sm = SpecManager()
-with mock.patch("biokbase.narrative.jobs.jobmanager.clients.get", get_mock_client):
-    sm.reload()
+with mock.patch(CLIENTS, get_mock_client):
+    sm = SpecManager()
     TEST_SPECS = copy.deepcopy(sm.app_specs)
-sm.reload()  # get live data
-LIVE_SPECS = copy.deepcopy(sm.app_specs)
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_job.py:59]: live specs are not used in this file, so these lines can be omitted

File: src/biokbase/narrative/tests/test_job.py [hunk 15]
```diff
@@ -203,34 +205,6 @@ def get_all_jobs(return_list=False):
     return jobs
 
 
-def get_cell_2_jobs():
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_job.py:206]: replicates what is in the `job_test_constants` file

File: src/biokbase/narrative/tests/test_updater.py [hunk 16]
```diff
@@ -14,31 +14,20 @@ def __init__(self, keyname, source):
 
 class UpdaterTestCase(unittest.TestCase):
     @classmethod
-    def setUpClass(self):
+    def setUpClass(cls):
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_updater.py:17]: these are minor flake8 / pylance errors that I fixed whilst in the neighbourhood

File: src/biokbase/narrative/tests/test_jobcomm.py [hunk 17]
```diff
@@ -3,21 +3,19 @@
 import os
 import itertools
 import re
+import copy
 
-from biokbase.narrative.exception_util import transform_job_exception
-from biokbase.narrative.jobs.jobcomm import exc_to_msg
-import biokbase.narrative.jobs.jobcomm
-import biokbase.narrative.jobs.jobmanager
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobcomm.py:10]: the imports had got a bit messed up, so I changed them to import the specific variables / classes / functions that are used

File: src/biokbase/narrative/tests/test_jobcomm.py [hunk 18]
```diff
@@ -131,41 +128,27 @@ def get_app_data(*args):
...
         config = ConfigTests()
         os.environ["KB_WORKSPACE_ID"] = config.get("jobs", "job_test_wsname")
-
-        cls.jc = biokbase.narrative.jobs.jobcomm.JobComm()
+        cls.jm = JobManager()
+        cls.jc = JobComm()
         cls.jc._comm = MockComm()
+        cls.job_states = get_test_job_states()
+        cls.retry_job = get_test_job_retries()
+        cls.job_info = get_test_job_infos()
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobcomm.py:145]: pre-generate all this data so the tests can just take the appropriate slice of the dictionary to get the expected results

File: src/biokbase/narrative/tests/test_jobcomm.py [hunk 19]
```diff
@@ -668,99 +651,55 @@ def test_lookup_job_states_by_cell_id__all_results(self):
     # -----------------------
     # Lookup job info
     # -----------------------
+
     @mock.patch(CLIENTS, get_mock_client)
-    def test_lookup_job_info__job_id__ok(self):
-        req_dict = make_comm_msg(INFO, {JOB_ID: JOB_COMPLETED}, False)
+    def check_job_info_results(self, job_args, job_id_list):
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobcomm.py:656]: generic checker for job_info requests

File: src/biokbase/narrative/tests/test_jobcomm.py [hunk 20]
```diff
@@ -870,23 +809,13 @@ def test_cancel_jobs__job_id_list__failure(self):
...
-    def test_retry_jobs__job_id__ok(self):
-        req_dict = make_comm_msg(RETRY, {JOB_ID: JOB_TERMINATED}, False)
-        expected = {
-            JOB_TERMINATED: {
-                "job_id": JOB_TERMINATED,
-                "job": self.job_states[JOB_TERMINATED],
-                "retry": get_retried_job(JOB_TERMINATED),
-            },
-        }
+    def check_retry_jobs(self, job_args, job_id_list):
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobcomm.py:814]: generic checker for retry results

File: src/biokbase/narrative/tests/test_jobcomm.py [hunk 21]
```diff
@@ -1116,25 +968,6 @@ def test_get_job_logs__job_id__no_job(self):
             self.jc._handle_comm_message(req_dict)
         self.check_error_message(req_dict, err)
 
-    @mock.patch(CLIENTS, get_mock_client)
-    def test_get_job_logs__job_id__bad_job(self):
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobcomm.py:1120]: this does the same thing as the test below -- checks a job ID that returns a not found error => deleted

File: src/biokbase/narrative/tests/test_jobmanager.py [hunk 22]
```diff
@@ -76,43 +79,20 @@
...
-
 __author__ = "Bill Riehl <wjriehl@lbl.gov>"
 
 
 TERMINAL_IDS = [JOB_COMPLETED, JOB_TERMINATED, JOB_ERROR]
 NON_TERMINAL_IDS = [JOB_CREATED, JOB_RUNNING]
-ERR_STR = "Some error occurred"
-
-
-def create_jm_message(r_type, job_id=None, data=None):
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobmanager.py:88]: unused

File: src/biokbase/narrative/tests/test_jobmanager.py [hunk 23]
```diff
@@ -136,26 +116,61 @@ def get_test_job_info(job_id):
     }
 
 
+def get_test_job_retries():
+    """
+    Expected output from a `retry_job` request
+    """
+    job_retries = {}
+    for job_id in TEST_JOB_IDS:
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobmanager.py:124]: realistic ee2 responses -- matching what the mock client does

File: src/biokbase/narrative/tests/test_jobmanager.py [hunk 24]
```diff
@@ -136,26 +116,61 @@ def get_test_job_info(job_id):
...
+        if job_id in RETRIED_JOBS:
+            job_retries[job_id] = {
+                "job_id": job_id,
+                "job": get_test_job_state(job_id),
+                # "retry_id": RETRIED_JOBS[job_id],
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobmanager.py:129]: DELETE ME!

File: src/biokbase/narrative/tests/test_jobmanager.py [hunk 25]
```diff
@@ -136,26 +116,61 @@ def get_test_job_info(job_id):
...
+        }
+
+    return job_retries
+
+
 class JobManagerTest(unittest.TestCase):
     @classmethod
     @mock.patch(CLIENTS, get_mock_client)
     def setUpClass(cls):
-        cls.job_ids = list(TEST_JOBS.keys())
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobmanager.py:143]: no need for this var as we already have this list as `TEST_JOB_IDS`

**Review by briehl — APPROVED** [review]
Much more streamlined!

**Pull request merged**